src/content/init.js

```diff
@@ -146,47 +146,56 @@ async function checkLoaded(settings) {
         pre.style.display = 'none';
     }
 
-    if (!loaded) {
-        flushData(settings);
-        loadedTimer = requestAnimationFrame(() =>
-            checkLoaded(settings).catch(rollbackPageChanges)
-        );
-        return;
-    }
-
     if (pre !== null) {
+        if (!loaded) {
+            flushData(settings);
+            loadedTimer = requestAnimationFrame(() =>
+                checkLoaded(settings).catch(rollbackPageChanges)
+            );
+            return;
+        }
+
         flushData(settings);
         pushChunk(null); // end of input
 
-        const [{ initDiscovery }, json] = await Promise.all([
-            import(chrome.runtime.getURL('discovery.js')),
-            data
-        ]);
-
-        await initDiscovery({
-            node: document.body,
-            raw: Object.defineProperties({}, {
-                firstSlice: {
-                    value: totalSize < firstSliceMaxSize * 2 ? null : firstSlice
-                },
-                size: {
-                    value: totalSize
-                },
-                json: totalSize <= firstSliceMaxSize ? { value: firstSlice } : {
-                    configurable: true,
-                    get() {
-                        return Object.defineProperty(this, 'json', {
-                            value: pre.textContent
-                        }).json;
+        const json = await data;
+
+        window.__discoveryPreloader = preloader; // eslint-disable-line no-underscore-dangle
```

Comment:
I suppose there is no need to expose preloader, since we just do `preloader.el.remove()` we can do it right before initDiscovery check

Reply:
We have to wait for the discovery initialization. In the case of the Fallback, this happens in a separate context, so I pass it through global. Tested it on large JSONs

Reply:
Well, preloader is needed while data is loading. Is it actual when data is loaded?

Reply:
Sure. Otherwise the boot will stop at data parsing step and you will see the Discovery immediately (skipping the init ui, prepare etc steps)